lib/ansible/modules/windows/win_toast.ps1

```diff
@@ -0,0 +1,101 @@
+#!powershell
+# This file is part of Ansible
+#
+# Copyright 2017, Jon Hawkesworth (@jhawkesworth) <figs@unity.demon.co.uk>
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+# WANT_JSON
+# POWERSHELL_COMMON
+
+$ErrorActionPreference = "Stop"
+
+# version check
+$osversion = [Environment]::OSVersion
+$lowest_version = 10
+if ($osversion.Version.Major -lt $lowest_version ) {
+   Fail-Json $result "Sorry, this version of windows, $osversion, does not support Toast notifications.  Toast notifications are available from version $lowest_version" 
+}
+
+$stopwatch = [system.diagnostics.stopwatch]::startNew()
+$now = [DateTime]::Now
+$default_title = "Notification: " + $now.ToShortTimeString()
+
+$params = Parse-Args $args -supports_check_mode $true
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
+
+# $expire_at = Get-AnsibleParam -obj $params -name "msg" -type "str"
+$expire_secs = Get-AnsibleParam -obj $params -name "expire_secs" -type "int" -default 45
+$group = Get-AnsibleParam -obj $params -name "group" -type "str" -default "Powershell"
+$msg = Get-AnsibleParam -obj $params -name "msg" -type "str" -default "Hello world!"
+$popup = Get-AnsibleParam -obj $params -name "popup" -type "bool" -default $true
+$tag = Get-AnsibleParam -obj $params -name "tag" -type "str" -default "Ansible"
+$title = Get-AnsibleParam -obj $params -name "title" -type "str" -default $default_title
+
+$timespan = New-TimeSpan -Seconds $expire_secs
+$expire_at = $now + $timespan
+$expire_at_utc = $($expire_at.ToUniversalTime()|Out-String).Trim()
+
+$result = @{
+    changed = $false
+    expire_secs = $expire_secs
+    expire_at = $expire_at.ToString()
+    expire_at_utc = $expire_at_utc
+    group = $group    
+    msg = $msg
```

Comment:
So I have asked myself this in the past as well, because I see various modules putting the input parameters also as result values. What is the purpose of this practice ? I don't see any, the input was provided by the user, so it should be the same as the result. Why would we register that stuff again ? Who would ever need it coming from the module.

In my opinion we should stop doing this, and return back to minimal return values that have a use (i.e. new output).

Also, this is already part of the invocation output anyway if you provide a higher verbosity.

Reply:
I suppose I do it because if the values are calculated as part of module params then you can use the calculated version from the registered results.

A contrived (and untested) example

win_toast:
  msg: "Warning {{release_name}} starts at {{ scheduled_release_time }}"
register: toast_message

# warn again
win_toast:
   msg: toast_message.result.msg

Reply:
I understand, but you already have the content of that message available to you. One could define it elsewhere (vars, set_fact) and use that variable twice.